src/mmap.rs

```diff
@@ -315,54 +320,53 @@ impl GuestMemoryMmap {
     /// Creates a container and allocates anonymous memory for guest memory regions.
     /// Valid memory regions are specified as a Vec of (Address, Size) tuples sorted by Address.
     pub fn new(ranges: &[(GuestAddress, usize)]) -> result::Result<Self, Error> {
-        if ranges.is_empty() {
-            return Err(Error::NoMemoryRegion);
-        }
-
-        let mut regions = Vec::<GuestRegionMmap>::new();
-        for range in ranges.iter() {
-            if let Some(last) = regions.last() {
-                if last
-                    .guest_base
-                    .checked_add(last.mapping.len() as GuestUsize)
-                    .map_or(true, |a| a > range.0)
-                {
-                    return Err(Error::MemoryRegionOverlap);
-                }
-            }
-
-            let mapping = MmapRegion::new(range.1).map_err(Error::MmapRegion)?;
-            regions.push(GuestRegionMmap {
-                mapping,
-                guest_base: range.0,
-            });
-        }
+        Self::with_files(ranges.iter().map(|r| (r.0, r.1, None)))
+    }
 
-        Ok(Self {
-            regions: Arc::new(regions),
-        })
+    /// Creates a container and allocates anonymous memory for guest memory regions.
+    /// Valid memory regions are specified as a Vec of (Address, Size, Option<FileOffset>)
+    /// tuples sorted by Address.
+    pub fn with_files<A, T>(ranges: T) -> result::Result<Self, Error>
+    where
+        A: Borrow<(GuestAddress, usize, Option<FileOffset>)>,
```

Comment:
Rust question: why is this Borrow trait needed here? What prevent us from using references in the code below?

Reply:
I used `Borrow` here because I wanted more flexibility in the `Item` the iterator returns. For example, if we would have stated the bound `T: IntoIterator<Item=&(GuestAddress, usize, Option<FileOffset>)>`, the current implementation of `new` would not work, because the `map` adapter creates an iterator that returns actual tuples (`(GuestAddress, usize, Option<FileOffset>)`) instead of references to them, and I can't simply add a `&` in front of the result because the compiler complains the reference might not live long enough. With `Borrow`, it works because we accept iterators whose items are anything which can be borrowed as `&(GuestAddress, usize, Option<FileOffset>)`.

`Borrow` is relatively similar to `AsRef` and `Deref`, but unlike them comes with the following blanket implementations: 
```rust
impl<'_, T> Borrow<T> for &'_ T
where
    T: ?Sized
```
and
```rust
impl<T> Borrow<T> for T
where
    T: ?Sized
```
which guarantee that both `(GuestAddress, usize, Option<FileOffset>)` and `&(GuestAddress, usize, Option<FileOffset>)` implement `Borrow<(GuestAddress, usize, Option<FileOffset>)`, and are thus usable as iterator item types in `with_files`.

Reply:
Makes sense, thanks for the explanation :)